### DISCUSSION TITLE
ReadImageToCanvas

### DISCUSSION THREAD

**Discussion by candcconsulting** [iTwin/itwinjs-core#3827] [Q&A] [answered]
In the sample here
[https://www.itwinjs.org/sandboxes/iTwinPlatform/Image%20Export](https://www.itwinjs.org/sandboxes/iTwinPlatform/Image%20Export)
shows how to export an image
For the current viewport we have the eyepoint set and the targetpoint is the center of the viewport.
However, we want to do an optical zoom
ie we are distanceValue away from the TargetPoint
`viewPort.zoom(undefined,distanceValue)`
but the zoom factor is not defined in units in
[https://www.itwinjs.org/reference/core-frontend/views/viewport/zoom/](https://www.itwinjs.org/reference/core-frontend/views/viewport/zoom/)
Then the real challenge is that the readImageToCanvas is dumping the image before it is fully rendered.
Is there a way to "await" this zoom ? (we have tried await)

**Comment by pmconne** [comment]
Please define "fully rendered". Is the problem that the view is animating while zooming and you're capturing the image before the animation completes, or that some contents of the view are not yet visible/refined, or something else?

**Reply by candcconsulting** [reply]
Thanks Paul,
There is geometry missing from the captured image (and this varies depending upon the scale factor etc and the status of the the tile count)
![viewport (1)](https://user-images.githubusercontent.com/19268085/174299299-957554fc-0f7c-4eb7-be51-981cabbdeb94.png)
![viewport (3)](https://user-images.githubusercontent.com/19268085/174299349-329ea2e1-0ca6-41ff-af77-da653f04531f.png)

**Comment by pmconne** [comment, marked as answer]
We can add an API to query whether a viewport's scene is complete. For the time being you could try something like:
```ts
async function waitForSceneCompletion(viewport: Viewport): Promise<void> {
  if (viewport.areAllTileTreesLoaded && viewport.numRequestedTiles === 0)
    return;

  await BeDuration.wait(100);
  return waitForSceneCompletion(viewport);
}

async function readImageToCanvas(viewport: Viewport): Promise<HTMLCanvasElement > {
  await waitForSceneCompletion(viewport);
  return viewport.readImageToCanvas();
}
```

**Reply by candcconsulting** [reply]
Thanks ... will try ...
This is what we have
```
  const takeSnapshot = async () => {
    const viewPort = IModelApp.viewManager.getFirstOpenView();
    if (viewPort !== undefined) {
      const clonedView = viewPort.view.clone();
      const zoom = document.getElementById('zoom') as HTMLInputElement;
      const returnZoomValue = await viewPort.zoom(undefined,parseFloat(zoom.value))
      console.log("Zoom Factor " + parseFloat(zoom.value) + " returned : " + returnZoomValue)
      const canvas = await viewPort.readImageToCanvas();
      const imageUrl = canvas.toDataURL("image/png");
      const link = document.createElement("a");
      const fileName = "zoom" + zoom.value + "png"
      link.setAttribute("download", fileName);
      link.setAttribute("href", imageUrl);
      link.click();
      viewPort.changeView(clonedView)
    }
  }
```
  will plug it in ... (right now all capturedImages are identical)

**Reply by pmconne** [reply]
FYI readImageToCanvas is synchronous - there's no need for you to `await` it.
Please let us know if issue #3828 accurately covers your requirements.

**Reply by candcconsulting** [reply]
Thanks ... it is "better" ... however, if the current viewPort is rendered then the BeDuration.wait is not hit ...
So if we takeSnapshot while the initial viewPort is tiling then it waits for the zoom to fully tile also
Also it would seem that the zoomFactor should be small if we want to zoom in ...
ie if I am 500m away from the target, but want to zoom so the target "fills" the viewport, is there a way to calculate the zoomFactor.
Note this has to be an optical zoom, the eyePoint cannot change we just need a zoomed in view of the targetPoint

**Reply by candcconsulting** [reply]
Sorry I just added a wait before checking waitForSceneCompletion and it works great.
Thanks

**Reply by candcconsulting** [reply]
the final working code
```
  async function waitForSceneCompletion(viewport: Viewport): Promise<void> {
    if (viewport.areAllTileTreesLoaded && viewport.numRequestedTiles === 0)
      return;
  
    await BeDuration.wait(100);
    return waitForSceneCompletion(viewport);
  }
  const takeSnapshot = async () => {
    const viewPort = IModelApp.viewManager.getFirstOpenView();
    if (viewPort !== undefined) {
      const clonedView = viewPort.view.clone();
      const zoom = document.getElementById('zoom') as HTMLInputElement;
      const returnZoomValue = viewPort.zoom(undefined,parseFloat(zoom.value))
      await BeDuration.wait(100);
      await waitForSceneCompletion(viewPort);
      console.log("Zoom Factor " + parseFloat(zoom.value) + " returned : " + returnZoomValue)
      const canvas = viewPort.readImageToCanvas();
      const imageUrl = canvas.toDataURL("image/png");
      const link = document.createElement("a");
      const fileName = "zoom" + zoom.value + ".png"
      link.setAttribute("download", fileName);
      link.setAttribute("href", imageUrl);
      link.click();
      viewPort.changeView(clonedView)
    }
  }
```

**Reply by candcconsulting** [reply]
PS : The fact it puts a Save in the Undo stack makes it very easy to step back and view the zoomed in view also

**Reply by pmconne** [reply]
> I just added a wait before checking waitForSceneCompletion

If you pass `{ animateFrustumChange: false }` as the third argument to `zoom` you shouldn't need this extra wait.

**Reply by candcconsulting** [reply]
Thanks ... and just for completeness the zoomFactor that works very nicely ... the attached was take 400m away from the signal
        zoomFactor = 1 / (distanceValue / 10);
![Targetx126069 276y262659 5023z17 2551Distance396 78518368397624](https://user-images.githubusercontent.com/19268085/174324304-988767b1-46b8-4ac0-a3b7-c355dbbb88f5.png)
